### PR TITLE
NETOBSERV-94: Create HPA for the console plugin

### PR DESCRIPTION
**Pull request by OlivierCazade**
Different things in this PR:
- fixed a bug in goflowkube HPA, the operator reconciler was always switching back the replicas number
- added HPA to the console plugin
- upgraded to HPA v2beta2
- some refactoring : splitted the plugin reconcile function due to cyclomatic complexity limit and added some constants

### CODE REVIEW

File: controllers/goflowkube/goflowkube_test.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	"gopkg.in/yaml.v2"
-	ascv1 "k8s.io/api/autoscaling/v1"
+	ascv2 "k8s.io/api/autoscaling/v2beta2"
```
[the remark this answers is not in the document]
Comment (jotak) [reply controllers/goflowkube/goflowkube_test.go:25]: Not in the upstream operator for sure, and even for downstream I would be surprised by such a restriction, but it's a good point to check

File: controllers/goflowkube/goflowkube_reconciler.go [hunk 2]
```diff
@@ -257,7 +257,10 @@ func autoScalerNeedsUpdate(asc *ascv1.HorizontalPodAutoscaler, desired *goflowKu
 	}
 	if asc.Spec.MaxReplicas != desired.HPA.MaxReplicas ||
 		differentPointerValues(asc.Spec.MinReplicas, desired.HPA.MinReplicas) ||
-		differentPointerValues(asc.Spec.TargetCPUUtilizationPercentage, desired.HPA.TargetCPUUtilizationPercentage) {
+		differentPointerValues(asc.Spec.MinReplicas, desired.HPA.MinReplicas) {
```
[the remark this answers is not in the document]
Comment (OlivierCazade) [reply controllers/goflowkube/goflowkube_reconciler.go:260]: Yes, thanks.

File: api/v1alpha1/flowcollector_types.go [hunk 3]
```diff
@@ -134,10 +135,9 @@ type FlowCollectorHPA struct {
 	MinReplicas *int32 `json:"minReplicas,omitempty" protobuf:"varint,2,opt,name=minReplicas"`
 	// upper limit for the number of pods that can be set by the autoscaler; cannot be smaller than MinReplicas.
 	MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,3,opt,name=maxReplicas"`
-	// target average CPU utilization (represented as a percentage of requested CPU) over all the pods;
-	// if not specified the default autoscaling policy will be used.
+	// Metrics used by the pod autoscaler
 	// +optional
-	TargetCPUUtilizationPercentage *int32 `json:"targetCPUUtilizationPercentage,omitempty" protobuf:"varint,4,opt,name=targetCPUUtilizationPercentage"`
+	Metrics []ascv2.MetricSpec `json:"metrics"`
```
[the remark this answers is not in the document]
Comment (OlivierCazade) [reply api/v1alpha1/flowcollector_types.go:140]: Yes, from here we can configure any metrics to scale with.

File: controllers/goflowkube/goflowkube_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (OlivierCazade) [reply controllers/goflowkube/goflowkube_test.go:25]: autoscaling v2 went GA in the last 1.23 k8s release. I first tried to use this and it worked well with openshift 4.10 but I had some issue with the controller-runtime envtest. It seems that autoscaling v2 was not enabled there despite having updated dependencies.

If this is a problem I can investigate more about this.

[the remark this answers is not in the document]
Comment (mariomac) [reply controllers/goflowkube/goflowkube_test.go:25]: Not a problem from my side

**Comment by jotak** [comment]
/hold
I prefer to hold this until we release v0.1.1

**Pull request merged**